### PR TITLE
Simplify shared preferences declaration with `sharedValue`

### PR DESCRIPTION
**Pull request by LunarX**

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/utils/SharedValue.kt [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+import android.content.SharedPreferences
+import com.infomaniak.lib.core.api.ApiController
+import com.infomaniak.lib.core.utils.Utils
+import com.infomaniak.lib.core.utils.transaction
+import kotlinx.serialization.encodeToString
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KProperty
+
+fun SharedPreferences.sharedValue(key: String, defaultValue: Boolean): ReadWriteProperty<Any, Boolean> {
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/utils/SharedValue.kt:28]: We should put this in Core

File: app/src/main/java/com/infomaniak/mail/utils/SharedValue.kt [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        override fun setValue(thisRef: Any, property: KProperty<*>, value: List<String>) {
+            transaction { putString(key, ApiController.json.encodeToString(value)) }
+        }
+    }
+}
+
+inline fun <reified E : Enum<E>> SharedPreferences.sharedValue(
+    key: String,
+    defaultValue: E
+): ReadWriteProperty<Any, E> {
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/utils/SharedValue.kt:92]: this can be onelined

**Review by FabianDevel — CHANGES_REQUESTED** [review]
I haven't finished yet

File: app/src/main/java/com/infomaniak/mail/utils/SharedValue.kt [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+fun SharedPreferences.sharedValue(key: String, defaultValue: List<String>): ReadWriteProperty<Any, List<String>> {
+    return object : ReadWriteProperty<Any, List<String>> {
+        override fun getValue(thisRef: Any, property: KProperty<*>): List<String> {
+            return getString(key, null)?.let(ApiController.json::decodeFromString) ?: defaultValue
+        }
+
+        override fun setValue(thisRef: Any, property: KProperty<*>, value: List<String>) {
+            transaction { putString(key, ApiController.json.encodeToString(value)) }
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/utils/SharedValue.kt:84]: I don't like that the json object we use here in in ApiController

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/utils/SharedValue.kt (hunk 3 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/utils/SharedValue.kt:84]: ApiController is just a static object holding properties, this `json` property is perfectly fine.

**Pull request merged**